src/components/sidebar/useCurrentActivePage.ts

```diff
@@ -0,0 +1,132 @@
+import { type FormPage } from '../../types';
+import { scrollIntoView } from '../../utils/form-helper';
+
+interface UseCurrentActivePageProps {
+  pages: FormPage[];
+  defaultPage: string;
+  activePages: string[];
+  evaluatedPagesVisibility: boolean;
+}
+
+interface UseCurrentActivePageResult {
+  currentActivePage: string | null;
+  requestPage: (pageId: string) => void;
+}
+
+/**
+ * Hook to manage the currently active page in a form sidebar.
+ *
+ * This implementation includes a locking mechanism to handle a specific limitation with Waypoint:
+ * When dealing with short forms where multiple pages are visible in the viewport simultaneously,
+ * Waypoint's initial visibility detection can be unpredictable. It might:
+ * 1. Report pages in a different order than their DOM position
+ * 2. Miss reporting some visible pages in the first few renders
+ * 3. Report visibility events before our desired initial scroll position is established
+ *
+ * The locking mechanism (isInitialPhaseRef) prevents these early Waypoint events from
+ * overriding our intended initial page selection. Without this lock:
+ * - The form might initially select the first page
+ * - But then immediately jump to a different page due to Waypoint's visibility events
+ * - This creates a jarring user experience where the form appears to "jump" during initialization
+ *
+ * The lock is released either:
+ * 1. Automatically after a timeout (allowing for initial render and scroll stabilization)
+ * 2. Immediately when the user explicitly interacts with the form
+ */
+export const useCurrentActivePage = ({
+  pages,
+  defaultPage,
+  activePages,
+  evaluatedPagesVisibility,
+}: UseCurrentActivePageProps): UseCurrentActivePageResult => {
+  const [currentActivePage, setCurrentActivePage] = useState<string | null>(null);
+  const [isInitialized, setIsInitialized] = useState(false);
+  const [requestedPage, setRequestedPage] = useState<string | null>(null);
+  // Use a ref to track if we're in the initial render phase
+  const initialLockTimeoutRef = useRef<NodeJS.Timeout | null>(null);
+  const isInitialPhaseRef = useRef(true);
+
+  // Initialize the active page
+  useEffect(() => {
+    if (isInitialized || !evaluatedPagesVisibility) return;
+
+    const initializePage = () => {
+      // Try to find and set the default page
+      const defaultPageObject = pages.find(({ label }) => label === defaultPage);
+
+      if (defaultPageObject && !defaultPageObject.isHidden) {
+        setCurrentActivePage(defaultPageObject.id);
+        scrollIntoView(defaultPageObject.id);
+      } else {
+        // Fall back to first visible page
+        const firstVisiblePage = pages.find((page) => !page.isHidden);
+        if (firstVisiblePage) {
+          setCurrentActivePage(firstVisiblePage.id);
+        }
+      }
+    };
+
+    initializePage();
+    setIsInitialized(true);
+  }, [pages, defaultPage, evaluatedPagesVisibility, isInitialized]);
```

Comment:
Why bother with the `useEffect()` here at all?:

```ts
if (!isInitialized && evaluatedPagesVisibility) {
  const defaultPageObject = pages.find(({ label }) => label === defaultPage);
  // blah, blah, blah
}
```

This is actually also a weird case where using a `ref` instead of state to track whether or not the hook is initialized is actually a performance improvement because the side effect of calling `setIsInitialized()` is actually undesirable (we only want the re-render if `setCurrentActivePage()` is called.

Reply:
Ah, yes. When I wrote this, I was hoping to get rid of most of the `useEffect()`s, but we definitely need it for the timeout